### ISSUE TITLE
Allow multiple callback services?

### ISSUE THREAD

**Issue by bonniegee** [RockefellerArchiveCenter/zodiac#147]
## Is your feature request related to a problem? Please describe.
Currently, a service can only have one callback service attached. Now that we're implementing Pictor, Ursa Major's Deliver Bags service ideally would have a callback to both Pictor's Bag Prepaper and Fornax's Assemble SIP.

## Describe the solution you'd like
Look into how complicated it would be to have multiple callbacks attached to a service.

## Describe alternatives you've considered
An alternative, and a short-term solution regardless, is to have a cron for the first Pictor service.

**Comment by helrond** [comment]
tl;dr this is not a simple fix, and we should set up crons in the short term.

Longer answer:
~~Because all the services are the same model, using Django many-to-many relationships doesn't work.~~

There are basically two kinds of services in Zodiac:
1. Services which do not require any external data to run. When triggered, these services usually look at database objects to figure out what needs to be processed. These services can process one or more objects at a time.
2. Services which act on POST data they received. These services always (I think) process only one object at a time.

The `ServiceRegistry` model currently has a `callback_service` field. The name of this field is a bit of a lie, because when the service completes it doesn't actually call the next service directly. Services are triggered by a shared task (`/gateway/tasks.py/queue_callbacks`) which looks for tasks which are another service's `callback_service` and then triggers a set number of those. What this means is that the main use of the `callback_service` field is creating the process visualization on the dashboard.

In very broad strokes, I think we have two options:
1. Abandon the "callback" facade we currently use, and more clearly delineate in our model which services can be triggered without supplying external data. Services which rely on their own internal state to run successfully can just be triggered on a regular basis, which is pretty much what's happening now. Systems which require POST data would not be triggered in these regular calls. If we went this route what we'd lose in the short term is our process visualization. There may be other ways to generate that though, and we could even consider removing it entirely. We'd also probably want to do something in the result view so that it's easier to filter out idle messages.
2. Lean into the idea of callbacks. This would require us to overhaul our architecture and go to a more stateless model where services are ALWAYS called with POST data, and thus ALWAYS operate on one thing at a time. The benefit to this would be that we'd have a single kind of service in our architecture, rather than a mix. It also means that the entire system would work a lot more efficiently, and we wouldn't have the situation we do now where a whole lot of idle messages are created. However, this is a major change, and would require changes to all of the microservice apps. I'm also not sure how well this approach would support troubleshooting when things go sideways.

Regardless of which option we choose, we'd want to provide some uniform method of validating POST data. With the second that would probably be more robust/complex, but even as things are we've run into a lot of confusion when data is passed between services. If we can define a what the shape of that data should be across all applications we can add a mixin to validate it in the views where it's expected.

**Comment by helrond** [comment]
Update, it looks like it _is_ actually possible to use a ManyToMany field for the purpose of adding multiple callbacks on the same model if the [symmetrical](https://docs.djangoproject.com/en/3.2/ref/models/fields/#django.db.models.ManyToManyField.symmetrical) argument is used. HOWEVER the larger questions of how meaningful these callback definitions really are remains.
